feat: notification for unread chats scrolled out of view

suprko opened the pull request:
#### Relevant info and issue/PR links 
https://app.clubhouse.io/peerio/story/838/desktop-unread-message-out-of-view-awareness
one text entry added to copy: https://github.com/PeerioTechnologies/peerio-icebear/pull/150

#### Testing instructions
Intended behav:
1. When there are unread messages scrolled out of view in ChatList, a little UI thingie shows up and points you towards unread messages, either at the top or bottom of ChatList depending on where the unreads are.
2. When there are unread messages above *and* below, this notification only shows up at the top. This is designers' decision, don't @ me.
3. Clicking on this notification causes ChatList to scroll until the next unread message is onscreen.

----
### Repository owner

- [ ] Was tested and can be merged.
- [ ] PR name follows conventional changelog format.

[PR guideline](https://github.com/PeerioTechnologies/peerio-icebear/blob/dev/docs/CONTRIBUTING.md)

suprko commented:
@anri-asaturov when you get a chance, I've got a couple of questions/concerns about this (see PR comment)

anri-asaturov commented:
Ok. I'm gonna review this now, but generally I'd prefer discussing this kind of things before implementation.

suprko commented on src/ui/chat/components/ChatList.jsx:
```diff
@@ -1,27 +1,287 @@
+    }
+
+    // Building the rooms & invites list
+    @computed get allRooms() {
```
Construction of the ChatList (rooms, room invites, and DMs) has been moved out of `render` to `computed` functions, which (I think?) is more performant but more importantly also allows them to be accessed by the other functions

suprko commented on src/ui/chat/components/ChatList.jsx:
```diff
@@ -40,47 +276,7 @@ class ChatList extends React.Component {
         routerStore.navigateTo(routerStore.ROUTES.newChannel);
     }
 
-    inviteOptions(kegDbId) {
```
All of these functions were removed, some were just leftover unused functions from previous ChannelInvites design, others are redundant one-liners that are better as in-line JSX.

suprko commented on src/ui/chat/components/ChatList.jsx:
```diff
@@ -22,6 +31,233 @@ class ChatList extends React.Component {
+        }
+    }
+
+    smoothScrollUp = () => {
```
"smoothScroll" functions borrowed from the one used in message list

suprko commented:
@anri-asaturov ready for review

1. Reworked so that it doesn't calculate any DOM `rect`s at all, just calculates ChatList `<li>` positions based on hardcoded padding & height values
2. Scroll-to-next/prev-unread on click works now
3. The unrelated edits (where I removed a bunch of unused functions) are still there. I marked that section in the comments.

anri-asaturov commented on src/ui/chat/components/ChatList.jsx:
```diff
@@ -22,6 +31,233 @@ class ChatList extends React.Component {
 
+    componentWillUnmount() {
+        window.removeEventListener('resize', this.checkScrollHeight);
+        this.scrollContainerRef.removeEventListener('scroll', this.checkScrollDistance);
```
Check for null ref just in case.

anri-asaturov commented on src/ui/chat/components/ChatList.jsx:
```diff
@@ -22,6 +31,233 @@ class ChatList extends React.Component {
+        this.checkScrollHeight();
+        window.addEventListener('resize', this.checkScrollHeight);
+
+        ref.addEventListener('scroll', this.checkScrollDistance);
```
Throttle this. There will be too many scroll events and each of them starts the whole recalculation.

anri-asaturov commented on src/ui/chat/components/ChatList.jsx:
```diff
@@ -22,6 +31,233 @@ class ChatList extends React.Component {
+        const positionsArray = [];
+
+        const unreadRooms = this.allRooms.reduce((acc, room, i) => {
+            if (room.unreadCount > 0) {
```
I think we should skip active chats, or the button will briefly show for them with every new message,

anri-asaturov commented on src/ui/chat/components/ChatList.jsx:
```diff
@@ -22,6 +31,233 @@ class ChatList extends React.Component {
+        });
+
+        const unreadDMs = chatStore.directMessages.reduce((acc, chat, i) => {
+            if (chat.unreadCount > 0) {
```
Here too - skip active chat

suprko commented:
@anri-asaturov all fixed

ebilg commented:
can be merged
